xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -1840,6 +1800,41 @@ void sendDiscoveryRequest(DiscoveryRequestData request) {
     void sendError(Exception error) {
       requestWriter.onError(error);
     }
+
+    @Override
+    public void onNext(DiscoveryResponse response) {
+      final DiscoveryResponseData responseData = DiscoveryResponseData.fromEnvoyProto(response);
+      if (logger.isLoggable(XdsLogLevel.DEBUG)) {
+        logger.log(XdsLogLevel.DEBUG, "Received {0} response:\n{1}",
+            responseData.getResourceType(), respPrinter.print(response));
+      }
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          handleDiscoveryResponse(responseData);
+        }
+      });
+    }
+
+    @Override
+    public void onError(final Throwable t) {
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          handleStreamClosed(Status.fromThrowable(t));
+        }
+      });
+    }
+
+    @Override
+    public void onCompleted() {
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          handleStreamClosed(Status.UNAVAILABLE.withDescription("Closed by server"));
+        }
+      });
+    }
```

Comment:
`stub.withWaitForReady().streamAggregatedResources(this)` looks bad. An extreme analogy of this PR:

```java
private final class AdsStreamV3 extends AbstractAdsStream
      implements StreamObserver<DiscoveryResponse>, Runnable {

  @Overrid
  public void run() {
    handleStreamClosed(Status.UNAVAILABLE.withDescription("Closed by server"));
  }

  @Override
  public void onCompleted() {
    syncContext.execute(this);
  }
}
```

Do you think this is clean and elegant?

Reply:
Wait, why would you write this code that is completely nonsense? You shouldn't write an arbitrary and meaningless code to demonstrate some approach is bad in general. 😢

Reply:
My point is AdsStream should be treated as composition not inheritance.

Reply:
AdsStream is not a response StreamObserver, just is it is not a Runnable in this extreme example. Response StreamObserver is only one of its components, a field of it at the most.

Reply:
Why an AdsStream cannot be a StreamObserver? See my comment at https://github.com/grpc/grpc-java/pull/7341#discussion_r473346296.  It can be a Runnable as well, but your code above just does not make sense. If it has some valid operations that needs to be scheduled, it's completely natural to make it a Runnable.